### ISSUE TITLE
Deadlock observed for -rc 2 and enc-mode 4

### ISSUE THREAD

**Issue by ttrigui** [OpenVisualCloud/SVT-AV1#819]
A deadlock in release mode happens when encoding with -rc 2 for -enc-mode 4 
-i F:\stream\ParkScene_1920x1080_24_60f.yuv  -rc 2  -intra-period 48 -fps-num 24000 -fps-denom 1001 -lp 8 -w 1920 -h 1080 -n 100 -b out1.ivf -enc-mode 4
The following tests did not fix the issue:
 -enable-altrefs 0
-fps 48 
increasing the size of  pa_reference_picture_buffer_init_count 
 -intra-period 47
 -intra-period 31

**Comment by JackZhouVCD** [comment]
I tracked the deadlock to be introduced by turning off the `NON_KF_INTRA_TF_FIX` macro (or this commit d061c4e). When the macro is turned on, a segmentation fault is seen with the following outputs:
```css
(gdb) file SvtAv1EncApp
Reading symbols from SvtAv1EncApp...done.
(gdb) run -i /home/ubuntu/stream/ParkScene_1920x1080_24_60f.yuv -intra-period 48 -fps-num 24000 -fps-denom 1001 -lp 8 -w 1920 -h 1080 -n 100 -enc-mode 4
Starting program: /home/ubuntu/Jack/SVT-AV1/Bin/Debug/SvtAv1EncApp -i /home/ubuntu/stream/ParkScene_1920x1080_24_60f.yuv -intra-period 48 -fps-num 24000 -fps-denom 1001 -lp 8 -w 1920 -h 1080 -n 100 -enc-mode 4
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
-------------------------------------------
SVT-AV1 Encoder
SVT [version]:  SVT-AV1 Encoder Lib v0.7.0
SVT [build]  :  GCC 7.4.0        64 bit
LIB Build date: Nov 25 2019 20:25:19
-------------------------------------------
Number of logical cores available: 8
Number of PPCS 72
-------------------------------------------
SVT [config]: Main Profile      Tier (auto)     Level (auto)
SVT [config]: EncoderMode                                                      : 4
SVT [config]: EncoderBitDepth / EncoderColorFormat / CompressedTenBitFormat    : 8 / 1 / 0
SVT [config]: SourceWidth / SourceHeight                                       : 1920 / 1080
SVT [config]: Fps_Numerator / Fps_Denominator / Gop Size / IntraRefreshType    : 24000 / 1001 / 49 / 1
SVT [config]: HierarchicalLevels / BaseLayerSwitchMode / PredStructure         : 4 / 0 / 2
SVT [config]: BRC Mode / QP  / LookaheadDistance / SceneChange                 : CQP / 50 / 33 / 0
-------------------------------------------
SVT Memory Usage:
    total allocated memory:       6.57 GB
        malloced memory:          4.99 GB
        callocated memory:        248.71 MB
        allocated aligned memory: 1.33 GB
    mutex count: 575
    semaphore count: 264
    thread count: 55
    hash table fulless: 0.110149, hash bucket is healthy
top 10 malloced memory locations:
(1.42 GB): /home/ubuntu/Jack/SVT-AV1/Source/Lib/Common/Codec/EbPictureControlSet.c:1230
(1.21 GB): /home/ubuntu/Jack/SVT-AV1/Source/Lib/Common/Codec/EbPictureControlSet.c:447
(310.78 MB): /home/ubuntu/Jack/SVT-AV1/Source/Lib/Common/Codec/EbCodingUnit.c:60
(292.76 MB): /home/ubuntu/Jack/SVT-AV1/Source/Lib/Common/Codec/EbPictureControlSet.c:98
(242.05 MB): /home/ubuntu/Jack/SVT-AV1/Source/Lib/Common/Codec/EbCodingUnit.c:61
(217.44 MB): /home/ubuntu/Jack/SVT-AV1/Source/Lib/Encoder/Codec/EbEncHandle.c:3443
(154.85 MB): /home/ubuntu/Jack/SVT-AV1/Source/Lib/Common/Codec/EbPictureControlSet.c:463
(143.05 MB): /home/ubuntu/Jack/SVT-AV1/Source/Lib/Common/Codec/EbBitstreamUnit.c:40
(138.16 MB): /home/ubuntu/Jack/SVT-AV1/Source/Lib/Common/Codec/EbModeDecisionProcess.c:254
(117.10 MB): /home/ubuntu/Jack/SVT-AV1/Source/Lib/Common/Codec/EbPictureControlSet.c:100
Encoding
Thread 13 "SvtAv1EncApp" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7ffe91d2c700 (LWP 14416)]
0x00007fffebb924ea in picture_decision_kernel (input_ptr=0x555584708ab0)
    at /home/ubuntu/Jack/SVT-AV1/Source/Lib/Common/Codec/EbPictureDecisionProcess.c:3808
3808                                        PictureParentControlSet* pcs_itr = (PictureParentControlSet*)encode_context_ptr->pre_assignment_buffer[pictureIndex - num_past_pics + pic_itr]->object_ptr;
```

**Comment by hassount** [comment]
this is related to #710

**Comment by hassount** [comment]
@JackZhouVCD can we try this again now that #870 is merged

**Comment by JackZhouVCD** [comment]
> @JackZhouVCD can we try this again now that #870 is merged

I just tried and I was unable to reproduce a deadlock in the current master, but neither could I get a deadlock in the commit right before. Taking into consideration that rc2 has been shifted to rc 1.

**Comment by hassount** [comment]
this should be addressed after a2ba299 and f50049a
let's close for now, and reopen if we find out otherwise
